proposals/flexible-vectors/README.md

```diff
@@ -1,7 +1,1184 @@
+- `vec.v16.m_store(m: vec.m16, a: memarg, v: vec.v16)`
+- `vec.v32.m_store(m: vec.m32, a: memarg, v: vec.v32)`
+- `vec.v64.m_store(m: vec.m64, a: memarg, v: vec.v64)`
+- `vec.v128.m_store(m: vec.m128, a: memarg, v: vec.v128)`
+
+## Lane operations
+
+### Splat scalar
+
+`idx` is interpreted modulo the length of the vector.
+
+- `vec.i8.splat(v: vec.v8, x: i32) -> vec.v8`
+- `vec.i16.splat(v: vec.v16, x: i32) -> vec.v16`
+- `vec.i32.splat(v: vec.v32, x: i32) -> vec.v32`
+- `vec.f32.splat(v: vec.v32, x: f32) -> vec.32`
+- `vec.i64.splat(v: vec.v64, x: i64) -> vec.v64`
+- `vec.f64.splat(v: vec.v64, x: f64) -> vec.v64`
+- `vec.v128.splat(v: vec.v128, x: v128) -> vec.v128`
+
+
+### Extract lane
+
+`idx` is interpreted modulo the length of the vector.
+
+- `vec.s8.extract_lane(v: vec.v8, idx: i32) -> i32`
+- `vec.u8.extract_lane(v: vec.v8, idx: i32) -> i32`
+- `vec.s16.extract_lane(v: vec.v16, idx: i32) -> i32`
+- `vec.u16.extract_lane(v: vec.v16, idx: i32) -> i32`
+- `vec.i32.extract_lane(v: vec.v32, idx: i32) -> i32`
+- `vec.f32.extract_lane(v: vec.v32, idx: i32) -> f32`
+- `vec.i64.extract_lane(v: vec.v64, idx: i32) -> i64`
+- `vec.f64.extract_lane(v: vec.v64, idx: i32) -> f64`
+- `vec.v128.extract_lane(v: vec.v128, idx: i32) -> v128`
+
+### Replace lane
+
+`idx` is interpreted modulo the length of the vector.
+
+- `vec.i8.extract_lane(v: vec.v8, idx: i32, x: i32) -> vec.v8`
+- `vec.i16.extract_lane(v: vec.v16, idx: i32, x: i32) -> vec.v16`
+- `vec.i32.extract_lane(v: vec.v32, idx: i32, x: i32) -> vec.v32`
+- `vec.f32.extract_lane(v: vec.v32, idx: i32, x: f32) -> vec.v32`
+- `vec.i64.extract_lane(v: vec.v64, idx: i32, x: i64) -> vec.v64`
+- `vec.f64.extract_lane(v: vec.v64, idx: i32, x: f64) -> vec.v64`
+- `vec.v128.extract_lane(v: vec.v128, idx: i32, x: v128) -> vec.v128`
+
+### Load lane
+
+Loads a single lane into existing vector.
+`idx` is interpreted modulo the length of the vector.
+
+- `vec.v8.load_lane(a: memarg, v: vec.v8, idx: i32) -> vec.v8`
+- `vec.v16.load_lane(a: memarg, v: vec.v16, idx: i32) -> vec.v16`
+- `vec.v32.load_lane(a: memarg, v: vec.v32, idx: i32) -> vec.v32`
+- `vec.v64.load_lane(a: memarg, v: vec.v64, idx: i32) -> vec.v64`
+- `vec.v128.load_lane(a: memarg, v: vec.v128, idx: i32) -> vec.v128`
+
+### Store lane
+
+Stores a single lane from vector
+`idx` is interpreted modulo the length of the vector.
+
+- `vec.v8.store_lane(a: memarg, v: vec.v8, idx: i32)`
+- `vec.v16.store_lane(a: memarg, v: vec.v16, idx: i32)`
+- `vec.v32.store_lane(a: memarg, v: vec.v32, idx: i32)`
+- `vec.v64.store_lane(a: memarg, v: vec.v64, idx: i32)`
+- `vec.v128.store_lane(a: memarg, v: vec.v128, idx: i32)`
+
+
+## Unary Arithmetic operators
+
+UNOP designates any unary operator (eg: neg, not)
+
+### UNOP
+
+- `vec.v8.UNOP(a: vec.v8) -> vec.v8`
+- `vec.v16.UNOP(a: vec.v16) -> vec.v16`
+- `vec.v32.UNOP(a: vec.v32) -> vec.v32`
+- `vec.v64.UNOP(a: vec.v64) -> vec.v64`
+- `vec.v128.UNOP(a: vec.v128) -> vec.v128`
```

Comment:
Actually, there are much more way to interpret a `vec.v128`. Basically, a `vec.v128` is a vector of `v128` (from fixed-size SIMD), so there are as many ways to interpret a `v128` than a `vec.v128`.

This section is a simplification of what the actual section will look, and the interpretation of the content of the vector will be part of the instruction. eg:

- `vec.i32.add(a: vec.v32) -> vec.v32`
- `vec.f32.add(a: vec.v32) -> vec.v32`

As element-wise unary and binary operation semantic is trivial, it has not been detailed in this proposal yet.

Reply:
OK, they do look kind of trivial, but they still raise some questions - for example, aren't the unmasked `vec.v128` operations completely redundant?

That's why I think it could be useful to state at least some of the masked operations concretely.

Reply:
I don't expect `vec.i8x16.add` to be a valid operation. The `vec.v128.UNOP` operations are for operations that deals with `v128` as a block, like reverse bytes. I'm not sure how many of those operations there will be.

Reply:
Oh, that's completely different; I was using the examples discussed in the description, i.e. `neg` and `not`. If you have such block use cases in mind, it's probably worth fleshing out the proposal a bit.